modules/4337/test/erc4337/ReferenceEntryPoint.spec.ts

```diff
@@ -215,6 +215,208 @@ describe('Safe4337Module - Reference EntryPoint', () => {
+    const signature = buildSignatureBytes([
+      {
+        signer: parentSafe.address,
+        data: await user.signTypedData(
+          {
+            verifyingContract: parentSafe.address,
+            chainId: await chainId(),
+          },
+          {
+            SafeMessage: [{ type: 'bytes', name: 'message' }],
+          },
+          {
+            message: opData,
+          },
+        ),
+        dynamic: true,
+      },
+    ])
+    const userOp = buildPackedUserOperationFromSafeUserOperation({
+      safeOp,
+      // Replace the 2nd word of static part of signature containing the pointer to dynamic part with invalid pointer value
+      signature: signature.slice(0, 67) + '00'.padStart(64, '0') + signature.slice(131),
+    })
+
+    await expect(entryPoint.handleOps([userOp], await relayer.getAddress()))
+      .to.be.revertedWithCustomError(entryPoint, 'FailedOp')
+      .withArgs(0, 'AA24 signature error')
+  })
+
+  it('should revert when expected offset and given dynamic pointer in signatures are not equal - Smart contract signature (NOTE: would require a staked paymaster for ERC-4337)', async () => {
+    const { user, relayer, safe: parentSafe, validator, entryPoint, safeGlobalConfig } = await setupTests()
+
+    await parentSafe.deploy(user)
+    const daughterSafe = Safe4337.withSigner(parentSafe.address, safeGlobalConfig)
+
+    const accountBalance = ethers.parseEther('1.0')
+    await user.sendTransaction({ to: daughterSafe.address, value: accountBalance })
+    expect(await ethers.provider.getBalance(daughterSafe.address)).to.be.eq(accountBalance)
+
+    const transfer = ethers.parseEther('0.1')
+    const safeOp = buildSafeUserOpTransaction(
+      daughterSafe.address,
+      user.address,
+      transfer,
+      '0x',
+      '0x0',
+      await entryPoint.getAddress(),
+      false,
+      false,
+      {
+        initCode: daughterSafe.getInitCode(),
+      },
+    )
+
+    const opData = calculateSafeOperationData(await validator.getAddress(), safeOp, await chainId())
+    const signature = buildSignatureBytes([
+      {
+        signer: parentSafe.address,
+        data: await user.signTypedData(
+          {
+            verifyingContract: parentSafe.address,
+            chainId: await chainId(),
+          },
+          {
+            SafeMessage: [{ type: 'bytes', name: 'message' }],
+          },
+          {
+            message: opData,
+          },
+        ),
+        dynamic: true,
+      },
+    ])
+
+    const userOp = buildPackedUserOperationFromSafeUserOperation({
+      safeOp,
+      // Replace the 2nd word of static part of signature containing the pointer to dynamic part with invalid pointer value
+      signature:
+        signature.slice(0, 66) +
+        ((signature.length - 2) / 2).toString(16).padStart(64, '0') +
+        '00' +
+        signature.slice(132) +
+        signature.slice(132),
+    })
```

Comment:
My personal opinion: It would be clearer to build the signature manually instead of first building it, then trying to change the signature to "make it work" for the test:

```suggestion
    const parentSafeSignature = user.signTypedData(
      {
        verifyingContract: parentSafe.address,
        chainId: await chainId(),
      },
      {
        SafeMessage: [{ type: 'bytes', name: 'message' }],
      },
      {
        message: opData,
      },
    )
    const signature = ethers.concat([
        ethers.zeroPadValue(parentSafe.address, 32), // address of the signer
        ethers.toHexBe(65 + "padding".length), // offset of the start of the signature
        "0x00", // contract signature type
        ethers.toUtf8Bytes("padding"), // extra illegal padding between signatures
        ethers.toHexBe(ethers.dataLength(parentSafeSignature)), // length of the dynamic signature
        parentSafeSignature,
    ])
```

(_disclaimer_: I didn't test the code out, but I found it more self-explanatory than all the slicing, especially since there are a lot of magic numbers there, since it is string slicing on `0x` prefixed hex instead of byte slicing).

Reply:
marking as resolved

https://github.com/safe-global/safe-modules/pull/453/commits/42b216e6f756cb43fe396f1947367fe0898a9584

https://github.com/safe-global/safe-modules/pull/453/commits/0306c399fa9cab12bb70279e56179cacc2cdaa26